### PR TITLE
Ensure GlobalActions are handled before anything else game-wide

### LINKED ISSUE #6026: Cannot exit overlays with M4
`GlobalAction.Back` doesn't work on Overlays when bound to mouse buttons (e.g. `M4` (mouse back button))
only when it's bound to keyboard keys (e.g. `Esc`)

Looks like key presses and mouse clicks are handled differently in general.
**osu!lazer version:** 
2019.906.0

### LINKED ISSUE #11914: Global bindings don't work when hovering certain drawables
In osu! we block mouse clicks from other elements on screen in multiple places by handling the `MouseDown` events:

https://github.com/ppy/osu/blob/45e41aaeacf930ce4b5e3b62f85acbc618ce73da/osu.Game/Screens/Select/Footer.cs#L104-L108

20210227 160026 (CleanShot)

This unfortunately blocks `GlobalAction`s from firing. An easy reproduction is to try and use the "back" binding (mouse 4 by default) while hovering the song select footer area.

I'm not immediately sure on the correct direction to fix this.

### PR DESCRIPTION
**Pull request by peppy**
Closes #11914.
Closes #6026.

This fails on some tests which require a bit more investigation (likely an issue with `ManualInputManager`) but I wanted to PR this to get some feedback on whether this is an acceptable direction, and to see if others can find edge cases where this introduces unexpected/unwanted behaviour.

In my initial testing it seems to work correctly in all cases.

### CODE REVIEW

File: osu.Game/Input/Bindings/GlobalActionContainer.cs [hunk 1]
```diff
@@ -14,11 +15,13 @@ public class GlobalActionContainer : DatabasedKeyBindingContainer<GlobalAction>,
     {
         private readonly Drawable handler;
 
-        public GlobalActionContainer(OsuGameBase game)
+        public GlobalActionContainer(OsuGameBase game, bool nested = false)
```
Comment (bdach) [inline osu.Game/Input/Bindings/GlobalActionContainer.cs:18]: The `nested` param doesn't seem to actually ever be read in any capacity.

File: osu.Game/Input/Bindings/GlobalActionContainer.cs [hunk 2]
```diff
@@ -14,11 +15,13 @@ public class GlobalActionContainer : DatabasedKeyBindingContainer<GlobalAction>,
...
             : base(matchingMode: KeyCombinationMatchingMode.Modifiers)
         {
             if (game is IKeyBindingHandler<GlobalAction>)
                 handler = game;
+
+            GetInputQueue = () => base.KeyBindingInputQueue.ToArray();
```
Comment (bdach) [inline osu.Game/Input/Bindings/GlobalActionContainer.cs:24]: What is the copy here for? Previous code doesn't seem to have needed one.

**Review by bdach — DISMISSED** [review]
Not sure overall on whether this is acceptable. It doesn't seem *bad*, just maybe a touch weird due to that `GetInputQueue` thing. It seems like it was once/was supposed to be driven by the `nested` param, but it's neither here or there.

File: osu.Game/Input/Bindings/GlobalActionContainer.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Input/Bindings/GlobalActionContainer.cs:18]: Remnants of a more messy PoC, can be removed.

File: osu.Game/Input/Bindings/GlobalActionContainer.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Input/Bindings/GlobalActionContainer.cs:24]: As above, unneeded.

File: osu.Game/Input/Bindings/GlobalActionContainer.cs [hunk 3]
```diff
@@ -12,11 +12,15 @@ namespace osu.Game.Input.Bindings
 {
     public class GlobalActionContainer : DatabasedKeyBindingContainer<GlobalAction>, IHandleGlobalKeyboardInput
     {
+        private readonly GlobalInputManager globalInputManager;
+
         private readonly Drawable handler;
 
-        public GlobalActionContainer(OsuGameBase game)
+        public GlobalActionContainer(OsuGameBase game, GlobalInputManager globalInputManager)
```
Comment (bdach) [inline osu.Game/Input/Bindings/GlobalActionContainer.cs:19]: Would stick a `[CanBeNull]` on the second param and the field.

**Comment by bdach** [comment]
As for the failure, the following seems to "fix" it:

```diff
diff --git a/osu.Game.Tests/Visual/Navigation/OsuGameTestScene.cs b/osu.Game.Tests/Visual/Navigation/OsuGameTestScene.cs
index bf5338d81a..372d50c003 100644
--- a/osu.Game.Tests/Visual/Navigation/OsuGameTestScene.cs
+++ b/osu.Game.Tests/Visual/Navigation/OsuGameTestScene.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
 using osu.Framework.Graphics;
@@ -13,6 +14,7 @@
 using osu.Game.Beatmaps;
 using osu.Game.Configuration;
 using osu.Game.Graphics.UserInterface;
+using osu.Game.Input.Bindings;
 using osu.Game.Online.API;
 using osu.Game.Overlays;
 using osu.Game.Rulesets;
@@ -65,6 +67,7 @@ public virtual void SetUpSteps()
             });
 
             AddUntilStep("Wait for load", () => Game.IsLoaded);
+            AddStep("Disable parent input", () => Game.ChildrenOfType<GlobalInputManager>().Single().UseParentInput = false);
             AddUntilStep("Wait for intro", () => Game.ScreenStack.CurrentScreen is IntroScreen);
 
             ConfirmAtMainMenu();
```

I haven't looked too long at whether that's a proper thing to do. What I can say for sure is that the bindings are getting handled twice - presumably because `GlobalInputManager` is in `OsuGameBase`, which the test browser derives from.

**Comment by peppy** [comment]
Right, the issue here is that `OsuGameTestScene` is creating its own local `GlobalActionContainer` which is getting precedence in handling. Due to the `PassThroughInputManager`'s presence, the global actions no longer propagate downwards (they are blocked by the next input manager by design).

While your solution does work around the issue, it's likely going to break if we ever test elements which require the actual fix in this PR (since it will be bypassing the new logic).

I think the correct way is to be able to disable the intermediary action handler in `OsuGameTestScene`, or remove the requirement of having it there. Will look into that.

**Review by smoogipoo — DISMISSED** [review]
I... think this is fine

**Comment by peppy** [comment]
Yeah, I'm not really 100% sure on this but I *am* sure that any issues that crop up will be noticed quite fast after a release.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**